crates/derive/src/lib.rs

```diff
@@ -0,0 +1,100 @@
+mod attr;
+
+mod derive_attr;
+mod derive_op;
+mod derive_shared;
+mod derive_type;
+
+use proc_macro::TokenStream;
+
+/// The `#[def_attribute]` proc macro.
+///
+/// The macro can be used to annotate a Rust struct as a new IR attribute.
+///
+/// The macro will leave the struct definition unchanged, but it will generate an implementation of
+/// the pliron::Attribute trait and implements other internal traits and types resources required
+/// to use the IR attribute.
+///
+/// Usage:
+///
+/// ```ignore
+/// #[def_attribute]
+/// #[attr_name = "my_dialect.attribute"]
+/// #[derive(Debug, Clone, PartialEq, Eq)]
+/// pub struct StringAttr(String);
+/// ```
+///
+/// **Note**: pre-requisite traits for `Attribute` must already be implemented.
+///         Additionaly, PartialEq must be implemented by the type.
+#[proc_macro_attribute]
+pub fn def_attribute(_args: TokenStream, input: TokenStream) -> TokenStream {
+    to_token_stream(derive_attr::def_attribute(input))
+}
+
+/// The `#[def_type]` proc macro.
+///
+/// The macro can be used to annotate a Rust struct as a new IR type.
+///
+/// The macro will leave the struct definition unchanged, but it will generate an implementation of
+/// the pliron::Type trait and implements other internal traits and types resources required
+/// to use the IR type.
+///
+/// Usage:
+///
+/// ```ignore
+/// #[def_type]
+/// #[attr_name = "my_dialect.unit"]
+/// #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+/// pub struct UnitType();
+/// ```
+///
+/// **Note**: pre-requisite traits for `Type` must already be implemented.
+///         Additionaly, Hash and Eq must be implemented by the rust type.
+#[proc_macro_attribute]
+pub fn def_type(_args: TokenStream, input: TokenStream) -> TokenStream {
+    to_token_stream(derive_type::def_type(input))
+}
+
+/// The `#[def_op]` proc macro.
+///
+/// The `#[def_op]` can be used to to create a new IR operation.
+///
+/// The macro assumes an empty struct and will add the `op: Ptr<Operation>` field used to access
+/// the underlying Operation in the context.
+///
+/// The macro will automatically derive the `Clone` and `Copy` traits for the new struct
+/// definition.
+///
+/// Usage:
+///
+/// ```ignore
```

Comment:
Is there any way to avoid having the user declare an empty `struct` ? In my design earlier, I had the user just provide the name. 
1. Having the user provide an empty struct, but later add an `op` field is kind of confusing for the user.
2. Are there checks to ensure that the user doesn't add extra fields? Ideally `Op` should just have that single field containing the `Operation` and nothing else. They're supposed to be lightweight wrappers around an `Operation`.

Reply:
Yeah, that part is not ideal. One could update the proc macro to accept a type-name and then create the struct for them. Similar to the original macro.

On the other hand I kinda like haveing the proc macro look like an annotation + having a `pub struct MyOp;`.

Plus IMO the doc strings like more natural using that style.

> Are there checks to ensure that the user doesn't add extra fields? Ideally Op should just have that single field containing the Operation and nothing else. They're supposed to be lightweight wrappers around an Operation.

Currently not, but it is easy to add checks.

Reply:
>On the other hand I kinda like haveing the proc macro look like an annotation + having a `pub struct MyOp;`.

Can we restrict the second part to *always* be `pub struct MyOp;` (exactly that, and not `pub struct MyOp {}` or `pub struct MyOp()` or anything else). In that case, I'm open to this syntax. I don't know if that's easy for you to implement, so if it isn't easy, let me know, we can discuss what to do.

>Plus IMO the doc strings like more natural using that style.

I agree.